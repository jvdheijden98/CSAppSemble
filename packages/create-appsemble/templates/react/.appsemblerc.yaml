actions:
  onLoad:
    required: true

output: dist
