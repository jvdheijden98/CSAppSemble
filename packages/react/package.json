{
  "name": "@appsemble/react",
  "version": "0.9.4",
  "dependencies": {
    "react-shadow-dom-retarget-events": "^1.0.10"
  },
  "peerDependencies": {
    "@appsemble/sdk": "^0.9.4",
    "react": "^16.11.0",
    "react-dom": "^16.11.0",
    "react-intl": "^3.4.0"
  }
}
