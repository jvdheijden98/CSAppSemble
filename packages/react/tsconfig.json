{
  "extends": "../../tsconfig",
  "exclude": [],
  "compilerOptions": {
    "lib": ["dom", "esnext"],
    "types": ["jest", "react-shadow-dom-retarget-events"]
  }
}
