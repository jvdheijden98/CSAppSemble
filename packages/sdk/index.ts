import {
  Action,
  BaseAction,
  Block,
  LinkAction,
  LogAction,
  Message,
  RequestAction,
  ResourceCreateAction,
  ResourceDeleteAction,
  ResourceGetAction,
  ResourceQueryAction,
  ResourceUpdateAction,
  Theme,
} from '@appsemble/types';
import { Promisable } from 'type-fest';

export {
  Action,
  BaseAction,
  LinkAction,
  LogAction,
  Message,
  RequestAction,
  ResourceCreateAction,
  ResourceDeleteAction,
  ResourceGetAction,
  ResourceQueryAction,
  ResourceUpdateAction,
  Theme,
};

export type Actions<A> = { [K in keyof A]: Action };

export interface PageParameters {
  [parameter: string]: string;
}

export interface Utils {
  /**
   * Register a function that should be called when the block is being removed.
   *
   * Use this to clean up resouces that would otherwise stay in memory, e.g. object URLs.
   */
  addCleanup: (fn: () => void) => void;

  /**
   * Show a bulma style message.
   */
  showMessage: (message: string | Message) => void;
}

export interface Events {
  /**
   * Emit an Appsemble event.
   *
   * @param type The type of event to emit.
   * @param data Data to emit with the event.
   */
  emit: (type: string | symbol, data: any) => void;

  /**
   * Remove an event listener for an Appsemble event.
   *
   * @param type The type of event to listen remove the listener from.
   * @param callback The callback to remove.
   */
  off: (type: string | symbol, callback: (event: Event) => any) => void;

  /**
   * Add an event listener for an Appsemble event.
   *
   * @param type The type of event to listen on.
   * @param callback A callback to register for the event.
   */
  on: (type: string | symbol, callback: (event: Event) => any) => void;
}

/**
 * The parameters that get passed to the bootstrap function.
 */
export interface BootstrapParams<P = any, A = {}> {
  /**
   * The actions that may be dispatched by the block.
   */
  actions: Actions<A>;

  /**
   * The block as it is defined in the app definition.
   */
  block: Block<P, A>;

  /**
   * Any kind of data that has been passed in by some context.
   */
  data: any;

  /**
   * Event related functions and constants.
   */
  events: Events;

  /**
   * URL parameters of the current route.
   *
   * If the page on which the block is rendered, has parameters specified on a page level, the
   * parameter keys and values of the page will be extracted and set as this object.
   */
  pageParameters?: PageParameters;

  /**
   * The theme of the block given the context.
   *
   * This is based on a combination of the base theme, app theme, page theme and block theme.
   */
  theme: Theme;

  /**
   * The shadow root to which DOM elements may be appended.
   */
  shadowRoot: ShadowRoot;

  /**
   * Some utility functions provided by the Appsemble app framework.
   */
  utils: Utils;
}

/**
 * @private
 */
export type BootstrapFunction<P = any, A = {}> = (
  params: BootstrapParams<P, A>,
) => Promisable<void>;

/**
 * @private
 */
export interface AppsembleBootstrapEvent extends CustomEvent {
  detail: {
    fn: BootstrapFunction;
    document: Document;
  };
}

/**
 * Register a boostrap function.
 *
 * @param fn The bootstrap function to register
 */
export function bootstrap<P = any, A = {}>(fn: BootstrapFunction<P, A>): void {
  const event = new CustomEvent('AppsembleBootstrap', {
    detail: {
      fn,
      document,
    },
  }) as AppsembleBootstrapEvent;
  if (document.currentScript) {
    document.currentScript.dispatchEvent(event);
  }
}

/**
 * Attach the returned node to the shadow root.
 *
 * This convenience wrapper attaches nodes returned by the bootstrap function to the shadow root.
 * This means that the initialization function for a block simply has to return a node.
 *
 * @param fn The bootstrap function to register.
 */
export function attach<P = any, A = {}>(
  fn: (params: BootstrapParams<P, A>) => Promisable<HTMLElement | void>,
): void {
  bootstrap<P, A>(
    async (params): Promise<void> => {
      const { shadowRoot } = params;

      const node = await fn(params);
      if (node instanceof HTMLElement) {
        shadowRoot.appendChild(node);
      }
    },
  );
}
