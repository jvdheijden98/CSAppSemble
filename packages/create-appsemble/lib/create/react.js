import fs from 'fs-extra';
import path from 'path';
import sortPackageJson from 'sort-package-json';

import { dependencies } from '../../templates/react/package.json';

export default async function react(outputPath, pkg) {
  const outPkg = sortPackageJson({
    ...pkg,
    dependencies: { ...pkg.dependencies, ...dependencies },
  });

  await fs.copy(path.resolve(__dirname, '../../templates/react'), outputPath);
  await fs.outputJSON(path.join(outputPath, 'package.json'), outPkg, { spaces: 2 });
}
