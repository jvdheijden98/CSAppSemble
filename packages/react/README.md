# Appsemble React SDK

> Build your own blocks using React

## Installation

```sh
npm install @appsemble/react
```

## Hello world example

```js
import { bootstrap } from '@appsemble/react';
import PropTypes from 'prop-types';
import React from 'react';

class MyBlock extends React.Component {
  static propTypes = {
    actions: PropTypes.shape().isRequired,
  };

  render() {
    const { actions } = this.props;

    return (
      <button onClick={actions.click.dispatch} type="button">
        Hello world!
      </button>
    );
  }
}

bootstrap(MyBlock);
```

**Note**: The script will be loaded only once. The component is then bootstrapped for every instance
of your block type that is loaded by the app creator.
