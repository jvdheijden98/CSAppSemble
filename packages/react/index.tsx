import { bootstrap as sdkBootstrap, BootstrapParams } from '@appsemble/sdk';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import retargetEvents from 'react-shadow-dom-retarget-events';

export interface BlockProps<P = any, A = {}> extends BootstrapParams<P, A> {
  /**
   * The DOM node on which the block is mounted.
   */
  reactRoot: HTMLElement;
}

const { Consumer, Provider } = React.createContext<BlockProps>(null);

/**
 * Mount a React component returned by a bootstrap function in the shadow DOM of a block.
 */
export function mount<P, A = {}>(
  Component: React.ComponentType<BlockProps<P, A>>,
  root?: HTMLElement,
): (params: BootstrapParams<P, A>) => void {
  return params => {
    const reactRoot = params.shadowRoot.appendChild(
      root ? root.cloneNode() : document.createElement('div'),
    ) as HTMLElement;
    const props = {
      ...params,
      reactRoot,
    };
    const component: JSX.Element = (
      <Provider value={props}>
        <Component {...props} />
      </Provider>
    );
    ReactDOM.render(component, reactRoot);
    params.utils.addCleanup(() => ReactDOM.unmountComponentAtNode(reactRoot));
    /**
     * React doesnâ€™t play nice with shadow DOM. This library works around that. However, the
     * implementation does contain some bugs.
     *
     * Related issues:
     *
     * - https://github.com/facebook/react/issues/2043
     * - https://github.com/facebook/react/issues/9242
     * - https://github.com/facebook/react/issues/10422
     * - https://github.com/facebook/react/issues/12973
     * - https://github.com/facebook/react/pull/8117
     * - https://github.com/facebook/react/pull/12163
     *
     * @todo Remove this when React has fixes this issue.
     */
    retargetEvents(params.shadowRoot);
  };
}

export function bootstrap<P, A = {}>(
  Component: React.ComponentType<BlockProps<P, A>>,
  reactRoot?: HTMLElement,
): void {
  sdkBootstrap<P, A>(mount(Component, reactRoot));
}

/**
 * A HOC which passes the Appsemble block values to he wrapped React component.
 */
export function withBlock<P extends object>(
  Component: React.ComponentType<P & Omit<BlockProps, keyof P>>,
): React.ComponentType<P> {
  function Wrapper(props: P): JSX.Element {
    return <Consumer>{values => <Component {...values} {...props} />}</Consumer>;
  }
  if (process.env.NODE_ENV !== 'production') {
    Wrapper.displayName = `withBlock(${Component.displayName || Component.name})`;
  }
  return Wrapper;
}
