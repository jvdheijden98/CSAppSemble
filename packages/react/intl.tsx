import * as React from 'react';
import { IntlConfig, IntlProvider } from 'react-intl';

import { BlockProps } from '.';

/**
 * A HOC which provides `intl` in a block written in React.
 */
// eslint-disable-next-line import/prefer-default-export
export function provideIntl<P = any, A = {}>(
  Component: React.ComponentType<BlockProps<P, A>>,
  intlProviderProps?: Partial<IntlConfig>,
): React.ComponentType<BlockProps<P, A>> {
  return props => (
    <IntlProvider defaultLocale="en-US" locale="en-US" {...intlProviderProps}>
      <Component {...props} />
    </IntlProvider>
  );
}
