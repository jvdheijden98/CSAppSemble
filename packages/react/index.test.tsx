import { BootstrapParams } from '@appsemble/sdk';
import * as enzyme from 'enzyme';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import retargetEvents from 'react-shadow-dom-retarget-events';

import { mount, withBlock } from '.';

jest.mock('react-dom', () => ({
  render: jest.fn(),
}));

jest.mock('react-shadow-dom-retarget-events', () => jest.fn());

let bootstrapParams: BootstrapParams;

beforeEach(() => {
  bootstrapParams = {
    actions: {},
    block: {
      type: 'test',
      version: '0.0.0',
      theme: {
        primaryColor: '#ff0000',
        linkColor: '#ff7f00',
        successColor: '#ffff00',
        infoColor: '#00ff00',
        warningColor: '#0000ff',
        dangerColor: '#4b0082',
        splashColor: '#8b00ff',
        themeColor: '#ffffff',
        tileLayer: 'https://cartodb-basemaps-c.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png',
      },
    },
    data: {},
    events: {
      emit() {},
      off() {},
      on() {},
    },
    shadowRoot: {
      appendChild(node: HTMLElement) {
        return node;
      },
    } as any,
    utils: {
      addCleanup() {},
      showMessage() {},
    },
    theme: {
      primaryColor: '#ff0000',
      linkColor: '#ff7f00',
      successColor: '#ffff00',
      infoColor: '#00ff00',
      warningColor: '#0000ff',
      dangerColor: '#4b0082',
      splashColor: '#8b00ff',
      themeColor: '#ffffff',
      tileLayer: 'https://cartodb-basemaps-c.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png',
    },
  };
});

describe('mount', () => {
  it('should mount the react component', () => {
    let component: JSX.Element;
    (ReactDOM.render as jest.Mock).mockImplementation(c => {
      component = c;
    });
    mount(() => <div>test</div>, document.createElement('div'))(bootstrapParams);
    expect(enzyme.mount(component)).toMatchSnapshot();
  });

  it('should retarget DOM events to the shadow root', () => {
    mount(() => <div>test</div>, document.createElement('div'))(bootstrapParams);
    expect(retargetEvents).toHaveBeenCalledWith(bootstrapParams.shadowRoot);
  });
});

describe('withBlock', () => {
  it('should pass block props to the child element', () => {
    const Test = withBlock(function Test() {
      return null;
    });
    let component: JSX.Element;
    (ReactDOM.render as jest.Mock).mockImplementation(c => {
      component = c;
    });
    mount(() => <Test />, document.createElement('div'))(bootstrapParams);
    expect(enzyme.mount(component)).toMatchSnapshot();
  });
});
