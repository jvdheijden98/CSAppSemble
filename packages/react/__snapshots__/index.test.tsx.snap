// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mount should mount the react component 1`] = `
<Component
  actions={Object {}}
  block={
    Object {
      "theme": Object {
        "dangerColor": "#4b0082",
        "infoColor": "#00ff00",
        "linkColor": "#ff7f00",
        "primaryColor": "#ff0000",
        "splashColor": "#8b00ff",
        "successColor": "#ffff00",
        "themeColor": "#ffffff",
        "tileLayer": "https://cartodb-basemaps-c.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png",
        "warningColor": "#0000ff",
      },
      "type": "test",
      "version": "0.0.0",
    }
  }
  data={Object {}}
  events={
    Object {
      "emit": [Function],
      "off": [Function],
      "on": [Function],
    }
  }
  reactRoot={<div />}
  shadowRoot={
    Object {
      "appendChild": [Function],
    }
  }
  theme={
    Object {
      "dangerColor": "#4b0082",
      "infoColor": "#00ff00",
      "linkColor": "#ff7f00",
      "primaryColor": "#ff0000",
      "splashColor": "#8b00ff",
      "successColor": "#ffff00",
      "themeColor": "#ffffff",
      "tileLayer": "https://cartodb-basemaps-c.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png",
      "warningColor": "#0000ff",
    }
  }
  utils={
    Object {
      "addCleanup": [Function],
      "showMessage": [Function],
    }
  }
>
  <div>
    test
  </div>
</Component>
`;

exports[`withBlock should pass block props to the child element 1`] = `
<Component
  actions={Object {}}
  block={
    Object {
      "theme": Object {
        "dangerColor": "#4b0082",
        "infoColor": "#00ff00",
        "linkColor": "#ff7f00",
        "primaryColor": "#ff0000",
        "splashColor": "#8b00ff",
        "successColor": "#ffff00",
        "themeColor": "#ffffff",
        "tileLayer": "https://cartodb-basemaps-c.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png",
        "warningColor": "#0000ff",
      },
      "type": "test",
      "version": "0.0.0",
    }
  }
  data={Object {}}
  events={
    Object {
      "emit": [Function],
      "off": [Function],
      "on": [Function],
    }
  }
  reactRoot={<div />}
  shadowRoot={
    Object {
      "appendChild": [Function],
    }
  }
  theme={
    Object {
      "dangerColor": "#4b0082",
      "infoColor": "#00ff00",
      "linkColor": "#ff7f00",
      "primaryColor": "#ff0000",
      "splashColor": "#8b00ff",
      "successColor": "#ffff00",
      "themeColor": "#ffffff",
      "tileLayer": "https://cartodb-basemaps-c.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png",
      "warningColor": "#0000ff",
    }
  }
  utils={
    Object {
      "addCleanup": [Function],
      "showMessage": [Function],
    }
  }
>
  <withBlock(Test)>
    <Test
      actions={Object {}}
      block={
        Object {
          "theme": Object {
            "dangerColor": "#4b0082",
            "infoColor": "#00ff00",
            "linkColor": "#ff7f00",
            "primaryColor": "#ff0000",
            "splashColor": "#8b00ff",
            "successColor": "#ffff00",
            "themeColor": "#ffffff",
            "tileLayer": "https://cartodb-basemaps-c.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png",
            "warningColor": "#0000ff",
          },
          "type": "test",
          "version": "0.0.0",
        }
      }
      data={Object {}}
      events={
        Object {
          "emit": [Function],
          "off": [Function],
          "on": [Function],
        }
      }
      reactRoot={<div />}
      shadowRoot={
        Object {
          "appendChild": [Function],
        }
      }
      theme={
        Object {
          "dangerColor": "#4b0082",
          "infoColor": "#00ff00",
          "linkColor": "#ff7f00",
          "primaryColor": "#ff0000",
          "splashColor": "#8b00ff",
          "successColor": "#ffff00",
          "themeColor": "#ffffff",
          "tileLayer": "https://cartodb-basemaps-c.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png",
          "warningColor": "#0000ff",
        }
      }
      utils={
        Object {
          "addCleanup": [Function],
          "showMessage": [Function],
        }
      }
    />
  </withBlock(Test)>
</Component>
`;
